[29-FEAT] 카카오 인증 기반 토큰 처리 구현

binimini opened the pull request:
## 작업 목록
- [x] 회원가입 로직 수정 (카카오 인증 서버 기반) 
- [x] 액세스 토큰 유효성 검증 로직 수정 (카카오 인증 서버 기반) 
- [x] 액세스 토큰 업데이트 로직 수정 (카카오 인증 서버 기반) 
- [x] 사용자 모델 수정 (리프레시 토큰 삭제, PK Auto Increment -> 카카오 회원번호로 변경)
- [x] 리프레시 토큰 기반 구현 삭제

## 세부 내용
- 사용자 모델의 `refreshToken` 컬럼을 삭제하고, PK를 카카오 인증서버에서 받아오는 회원번호 기반으로 수정했습니다. 
  (`Auto Increment` 속성 삭제)
 DB 변경문은 아래로 적용하시면 됩니당~
  ```
  alter table User drop column refreshToken;
  set foreign_key_checks = 0;
  alter table User modify userId int not null;
  set foreign_key_checks = 1;
  ```

- 사용자 정보를 바디와 헤더(토큰)로 받아서 회원가입을 하던 방식을 Bearer 방식의 액세스 토큰만 받고
  카카오 인증 서버에서 사용자 정보를 가져와 회원가입하는 방식으로 변경했습니다!

- 에러 핸들러를 적용안한 상태로 개발해서 머지한 뒤에 에러 처리하는 방식 추가 수정하겠습니당~ 
- 토큰 관련 추가 구현 완료했습니다! 카카오 인증 서버에서 액세스 토큰 받아서 테스트해보려고 했는데 
 REST API 액세스 토큰 받는 과정에서 요청이 안들어와서 🤣🤣 먼저 올리고 추가적으로 테스트해보도록 하겠습니당..! 
- 받는 요청 디버깅 어려워서 Morgan 일단 깔아서 요청 들어오는 것만 보이도록 했습니다. 추가적인 로깅은 찬찬히 해봐욥 🙃
  <img width="654" alt="스크린샷 2022-06-20 오후 9 34 03" src="https://user-images.githubusercontent.com/69030160/174603004-ed51bf12-822b-4854-b8e3-548e3fd717ca.png">

## 논의 사항
- 없음

## 연결된 이슈
- #29

binimini committed:
fix: 에러 픽스

- 컴마 제거 후 prettier 적용

binimini committed:
refactor: Service 단에서의 try/catch문 제거

binimini committed:
feat: Auth 관련 추가 구현

- 토큰 갱신 함수 UserService 내 추가
- AuthController 구현
- Middlware 네이밍 및 DTO 수정

binimini committed:
chore: Axios 의존성 추가

- OAuth 카카오 토큰 관련 요청 클라이언트로 Axios 사용

binimini committed:
feat: 카카오 인증 관련 요청 로직 구현

- AuthService로 분리
- 추후 서드파티 에러 처리 필요
- userId AuthService 통해 받은후
- UserService 통해 update (token 갱신된 경우)

binimini committed:
feat: 리프레시 토큰 기반 구현 변경 및 삭제

- AuthController 삭제
- 카카오 인증서버 기반으로 수정

binimini committed:
refactor: 필요없는 코드 삭제

- 사용하지않는 DTO 삭제

binimini committed:
chore: Morgan 의존성 추가

- 요청 트래픽 확인할 수 있도록 로거 의존성 추가

binimini committed:
feat: 토큰 파싱시 Bearer 검증 추가

cpplovelove commented on app.ts:
```diff
@@ -4,9 +4,13 @@ import UserController from './controllers/user-controller';
 
 const app = express();
 
+const LOGGER = process.env.LOGGER || 'dev';
```
로그 찍어주는 거 좋은 것 같아요! 다른 로그 라이브러리도 있는 지 찾아보겠습니다!

cpplovelove commented on services/promising-service.ts:
```diff
@@ -1,27 +1,23 @@
-      return new InternalServerError('Internal Server Error');
-    }
+    promisingInfo.category = category;
+    promisingInfo = await arrayUtil.deleteJsonKey(promisingInfo, 'categoryId');
```
엇 이 부분 코드 리뷰해주신거 참조해서 저도 고쳐서 PR에 푸시해놨는데 더 깔끔하게 적어주신 것 같아서 머지하면서 제 부분 없애놓도록 하겠습니다!

binimini replied:
로직 자체는 똑같아서 아무거나 사용해도 괜찮을 거 같아용 좋아용~

cpplovelove reviewed (commented):
너무 수고많으셨습니다! ☘️
작성해주신 db변경 쿼리문 rds에 적용시켜놨습니다! 
요청들어오는 로그 morgan으로 찍는 거 좋은 것 같습니다! 코드 보면서 생각난건데 에러 핸들러 작성하면서 저희 로그 찍을 때 winston라이브러리 사용하면 어떨지 여쭤보려했거든요! 혹시 어떠신가요? 멋진 로그를 찍을 수 있습니다 .. https://github.com/winstonjs/winston

binimini commented:
맞아요 winston도 괜찮은 거 같은데 morgan이랑 결합해서 쓰기도 하는 것 같더라고요?
그래서 공부하면서 차차 적용해보려구 했습니당 쓰는 거 좋아요! 😀

cpplovelove commented:
좋아요! 따로 수정 필요한 부분은 없는 것 같습니다! 머지해도 될 것 같아요!

binimini committed:
feat: 에러 핸들러 미들웨어 추가 및 설정

- 컨플릭트 해결
- 공통 미들웨어로 설정 (next(err)로 에러 핸들러로 넘기기)
- ApplicationError 처리
- 이외 에러 ApplicationError로 변환해 처리

binimini commented:
선영님 커스텀 에러는 구현되어있는데 전체적인 에러 잡아주는 공통 미들웨어가 없길래 추가해놨어요!
이제 서비스처럼 아래에서 불리는 경우에는 에러 던져주고
컨트롤러나 미들웨어처럼 바깥에서 캐치해주는 레이어가 없는 경우에는 next(err)로 감싸주면 에러 핸들러로 처리됩니다 ✨
```ts
@JsonController()
class Controller {
  async func(@Req() req: Request, @Res() res: Response, next: NextFunction) {
    try {
     // 서비스 호출 로직
    } catch (err: any) {
      next(err);
    }
  }
}
```
결과는 이런식으로 나옵니다~
<img width="1010" alt="스크린샷 2022-06-21 오후 7 52 00" src="https://user-images.githubusercontent.com/69030160/174785135-e2101bae-db6d-486c-a0f3-e9bea8683fad.png">
